File: supalink/src/main/java/app/kyulabs/supalink/SupalinkSDK.java
package app.kyulabs.supalink;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class SupalinkSDK {

    public interface CompletionHandler {
        void onComplete(Supalink supalink);
    }

    public static void handleAsestLinks(Context context, Intent intent, CompletionHandler completionHandler) {
        if (intent != null && intent.getData() != null) {
            Uri data = intent.getData();
            Supalink supalink = urlToSupalink(data);
            if (supalink != null) {
                if (supalink.getAfl() != null) {
                    Uri uri = Uri.parse(supalink.getAfl());
                    if (canOpenUrl(context, uri)) {
                        openUrl(context, uri);
                    }
                }
                completionHandler.onComplete(supalink);
            } else {
                completionHandler.onComplete(null);
            }
        } else {
            completionHandler.onComplete(null);
        }
    }

    private static Supalink urlToSupalink(Uri url) {
        Map<String, String> linkValues = new HashMap<>();

        Set<String> queryParameterNames = url.getQueryParameterNames();
        for (String key : queryParameterNames) {
            String value = url.getQueryParameter(key);
            if (value != null) {
                linkValues.put(key, value);
            }
        }

        if (linkValues.isEmpty()) {
            return null;
        }

        return new Supalink(
                linkValues.get("link"),
                linkValues.get("apn"),
                linkValues.get("afl"),
                linkValues.get("amv"),
                linkValues.get("ibi"),
                linkValues.get("ifl"),
                linkValues.get("ius"),
                linkValues.get("ipfl"),
                linkValues.get("ipbi"),
                linkValues.get("isi"),
                linkValues.get("imv"),
                linkValues.get("eft"),
                linkValues.get("ofl"),
                linkValues.get("st"),
                linkValues.get("sd"),
                linkValues.get("si"),
                linkValues.get("utm_source"),
                linkValues.get("utm_medium"),
                linkValues.get("utm_campaign"),
                linkValues.get("utm_term"),
                linkValues.get("utm_content"),
                linkValues.get("at"),
                linkValues.get("ct"),
                linkValues.get("mt"),
                linkValues.get("pt")
        );
    }

    private static boolean canOpenUrl(Context context, Uri url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, url);
        return intent.resolveActivity(context.getPackageManager()) != null;
    }

    private static void openUrl(Context context, Uri url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, url);

        // Context가 Activity일 경우만 startActivity 호출
        if (context instanceof Activity) {
            context.startActivity(intent);
        } else {
            // ApplicationContext라면 적절한 처리를 추가할 수 있습니다.
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        }
    }
}

File: Supalink/src/main/java/app/kyulabs/supalink/SupalinkSDK.kt
package app.kyulabs.supalink
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.ContextCompat.startActivity

/**
 * Supalink SDK 클래스
 */
object SupalinkSDK {
    // 수정된 타입
    fun handleAsestLinks(context: Context, intent: Intent?, completionHandler: (Supalink?) -> Unit) {
        intent?.data?.let {
            urlToSupalink(it)?.let { supalink ->
                if (supalink.afl != null) {
                    val uri = Uri.parse(supalink.afl)
                    if (canOpenUrl(context, uri)) {
                        openUrl(context, uri)
                    }
                }

                completionHandler(supalink)
            }
        } ?: completionHandler(null)
    }

    private fun urlToSupalink(url: Uri): Supalink? {
        val linkValues: MutableMap<String, String> = mutableMapOf()

        url.queryParameterNames.forEach { key ->
            url.getQueryParameter(key)?.let {
                linkValues[key] = it
            }
        }

        return if (linkValues.isEmpty()) null else Supalink(
            link = linkValues["link"],
            apn = linkValues["apn"],
            afl = linkValues["afl"],
            amv = linkValues["amv"],
            ibi = linkValues["ibi"],
            ifl = linkValues["ifl"],
            ius = linkValues["ius"],
            ipfl = linkValues["ipfl"],
            ipbi = linkValues["ipbi"],
            isi = linkValues["isi"],
            imv = linkValues["imv"],
            eft = linkValues["eft"],
            ofl = linkValues["ofl"],
            st = linkValues["st"],
            sd = linkValues["sd"],
            si = linkValues["si"],
            utm_source = linkValues["utm_source"],
            utm_medium = linkValues["utm_medium"],
            utm_campaign = linkValues["utm_campaign"],
            utm_term = linkValues["utm_term"],
            utm_content = linkValues["utm_content"],
            at = linkValues["at"],
            ct = linkValues["ct"],
            mt = linkValues["mt"],
            pt = linkValues["pt"]
        )
    }

    private fun canOpenUrl(context: Context, url: Uri): Boolean {
        val intent = Intent(Intent.ACTION_VIEW, url)
        return intent.resolveActivity(context.packageManager) != null
    }

    private fun openUrl(context: Context, url: Uri) {
        val intent = Intent(Intent.ACTION_VIEW, url)

        // Context가 Activity일 경우만 startActivity 호출
        if (context is Activity) {
            context.startActivity(intent)
        } else {
            // ApplicationContext라면 적절한 처리를 추가할 수 있습니다.
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }
    }
}
